mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java

```diff
@@ -880,15 +866,19 @@ public void onData(ByteBuffer data, boolean endStream, EnvoyStreamIntel streamIn
         return;
       }
 
+      ByteBuffer userBuffer = mUserCurrentReadBuffer;
+      mUserCurrentReadBuffer = null; // Avoid the reference to a potentially large buffer.
+      int dataRead = data.remaining();
+      // It is important to copy the `data` into the `userBuffer` outside the thread execution
+      // because the `data` is backed by a direct `ByteBuffer` and it will be destroyed upon once
```

Comment:
nit: s/destroyed upon once/destroyed once

Also, was this a problem before this PR?  If not, why is it a problem now?  I think I'm missing the correlation between this particular change and the rest of the PR

Reply:
> nit: s/destroyed upon once/destroyed once

Fixed the typo.

> Also, was this a problem before this PR? If not, why is it a problem now? I think I'm missing the correlation between this particular change and the rest of the PR

With the new change, we are passing the direct `ByteBuffer` into the callback without any copy (no more `envoy_data` (in C++) to `byte[]` (in Java). The direct `ByteBuffer` is backed by the native C++ `Envoy::Buffer::Instance`. That also means it's tied to the lifetime of `Envoy::Buffer::Instance`. In Cronvoy, we run the `onData` on a separate thread, so we have to make a copy of the `ByteBuffer`, which is what the existing code does (the old is also making a copy). The new code is simply moving the copy outside the thread because making a copy inside the thread poses a risk of having a dangling `ByteBuffer`.